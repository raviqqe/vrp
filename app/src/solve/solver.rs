pub trait SimpleSolver: BasicSolver<Vehicle, Stop, SimpleProblem, Solution> {}
