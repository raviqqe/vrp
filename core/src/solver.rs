use crate::{problem::BasicProblem, BasicSolution, BasicStop, BasicVehicle};

pub trait BasicSolver<V: BasicVehicle, S: BasicStop, P: BasicProblem<V, S>, L: BasicSolution> {
    fn solve(&mut self, problem: &P) -> L;
}
