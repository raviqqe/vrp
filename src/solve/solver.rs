use crate::Problem;

pub trait Solver {
    // TODO Reason no solution.
    fn solve(&self, problem: &Problem) -> Option<Problem>;
}
