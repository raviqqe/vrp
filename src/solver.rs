pub trait Solver {
    fn solve() -> foo {}
}
