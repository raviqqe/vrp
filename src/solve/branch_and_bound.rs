use super::solver::Solver;
use crate::{cost::CostCalculator, hash_map::HashMap, problem::BaseProblem, Solution};
use bumpalo::Bump;
use ordered_float::OrderedFloat;

pub struct BranchAndBoundSolver<C: CostCalculator> {
    cost_calculator: C,
}

impl<C: CostCalculator> BranchAndBoundSolver<C> {
    pub fn new(cost_calculator: C) -> Self {
        Self { cost_calculator }
    }
}

impl<C: CostCalculator> Solver for BranchAndBoundSolver<C> {
    fn solve(&mut self, problem: impl BaseProblem) -> Solution {
        let allocator = Bump::new();
        let mut solutions = HashMap::default();
        let solution = Solution::new({
            let mut routes = Vec::with_capacity_in(problem.vehicle_count(), &allocator);
            routes.extend((0..problem.vehicle_count()).map(|_| Vec::new_in(&allocator)));
            routes
        });
        let cost = self.cost_calculator.calculate(&solution);
        solutions.insert(solution, cost);
        let mut new_solutions = vec![];

        for stop_index in 0..problem.stop_count() {
            new_solutions.clear();

            for (solution, upper_bound) in &solutions {
                for vehicle_index in 0..solution.routes().len() {
                    let solution = solution.add_stop(vehicle_index, stop_index);
                    let lower_bound = self.cost_calculator.calculate_lower_bound(&solution);

                    if lower_bound < *upper_bound {
                        let cost = self.cost_calculator.calculate(&solution);
                        new_solutions.push((solution, cost));
                    }
                }
            }

            solutions.extend(new_solutions.drain(..));
        }

        let solution = solutions
            .into_iter()
            .min_by(|(_, one), (_, other)| OrderedFloat(*one).cmp(&OrderedFloat(*other)))
            .expect("at least one solution")
            .0;

        solution.to_global()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        cost::{DeliveryCostCalculator, DistanceCostCalculator},
        Location, SimpleProblem, Stop, Vehicle,
    };
    use insta::assert_debug_snapshot;

    const DISTANCE_COST: f64 = 1.0;
    const MISSED_DELIVERY_COST: f64 = 1e9;

    fn solve(problem: &SimpleProblem) -> Solution {
        BranchAndBoundSolver::new(DeliveryCostCalculator::new(
            DistanceCostCalculator::new(problem),
            problem.stops().len(),
            MISSED_DELIVERY_COST,
            DISTANCE_COST,
        ))
        .solve(problem)
    }

    #[test]
    fn do_nothing() {
        let problem = SimpleProblem::new(vec![Vehicle::new()], vec![]);

        assert_eq!(solve(&problem), Solution::new(vec![vec![]]));
    }

    #[test]
    fn keep_one_stop() {
        let problem = SimpleProblem::new(
            vec![Vehicle::new()],
            vec![Stop::new(Location::new(0.0, 0.0))],
        );

        assert_eq!(solve(&problem), Solution::new(vec![vec![0]]));
    }

    #[test]
    fn keep_two_stops() {
        let problem = SimpleProblem::new(
            vec![Vehicle::new()],
            vec![
                Stop::new(Location::new(0.0, 0.0)),
                Stop::new(Location::new(1.0, 0.0)),
            ],
        );

        assert_debug_snapshot!(solve(&problem));
    }

    #[test]
    fn keep_three_stops() {
        let problem = SimpleProblem::new(
            vec![Vehicle::new()],
            vec![
                Stop::new(Location::new(0.0, 0.0)),
                Stop::new(Location::new(1.0, 0.0)),
                Stop::new(Location::new(2.0, 0.0)),
            ],
        );

        assert_debug_snapshot!(solve(&problem));
    }

    #[test]
    fn optimize_stop_order() {
        let problem = SimpleProblem::new(
            vec![Vehicle::new()],
            vec![
                Stop::new(Location::new(0.0, 0.0)),
                Stop::new(Location::new(2.0, 0.0)),
                Stop::new(Location::new(1.0, 0.0)),
            ],
        );

        assert_debug_snapshot!(solve(&problem));
    }

    #[test]
    fn even_workload() {
        let problem = SimpleProblem::new(
            vec![Vehicle::new(), Vehicle::new()],
            vec![
                Stop::new(Location::new(0.0, 0.0)),
                Stop::new(Location::new(1.0, 0.0)),
                Stop::new(Location::new(2.0, 0.0)),
            ],
        );

        assert!(solve(&problem).routes().iter().all(|stops| stops.len() < 3));
    }
}
