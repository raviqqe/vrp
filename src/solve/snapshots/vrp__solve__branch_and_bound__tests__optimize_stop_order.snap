---
source: src/solve/branch_and_bound.rs
expression: solve(&problem)
---
Solution {
    routes: [
        [
            1,
            2,
            0,
        ],
    ],
}
